dev/com.ibm.ws.security.jwt/src/com/ibm/ws/security/jwt/internal/ConsumerUtil.java

```diff
@@ -58,914 +59,918 @@
+			String errorMsg = Tr.formatMessage(tc, "JWE_REQUIRED_BUT_TOKEN_NOT_JWE", new Object[] { config.getId() });
+			throw new InvalidTokenException(errorMsg);
+		}
+	}
+
+	JwtContext getJwtContextFromCache(@Sensitive String jwtString, JwtConsumerConfig config) {
+		initializeCache();
+		return (JwtContext) jwtCache.get(jwtString, config.getId());
+	}
+
+	private synchronized void initializeCache() {
+		long timeoutMillis = 1000 * 60 * 5;
+		if (jwtCache == null) {
+			jwtCache = new JwtCache(timeoutMillis);
+		}
+	}
+
+	void cacheJwtContext(@Sensitive String jwtString, JwtContext jwtContext, JwtConsumerConfig config) {
+		initializeCache();
+		jwtCache.put(jwtString, config.getId(), jwtContext, config.getClockSkew());
+	}
+
+	JwtContext parseNewJwtWithoutValidation(@Sensitive String jwtString, JwtConsumerConfig config)
+			throws InvalidTokenException, InvalidJwtException {
+		if (JweHelper.isJwe(jwtString)) {
+			jwtString = JweHelper.extractJwsFromJweToken(jwtString, config, mpConfigProps);
+		}
+		JwtConsumerBuilder builder = initializeJwtConsumerBuilderWithoutValidation(config);
+		JwtConsumer firstPassJwtConsumer = builder.build();
+		return firstPassJwtConsumer.process(jwtString);
+	}
+
+	protected void validateJwtContext(JwtContext jwtContext, String jwtString, JwtConsumerConfig config)
+			throws Exception {
+		Key key = getSigningKey(config, jwtContext);
+		JwtClaims jwtClaims = jwtContext.getJwtClaims();
+
+		if (tc.isDebugEnabled()) {
+			Tr.debug(tc, "Key from config: " + key);
+		}
+
+		validateClaims(jwtClaims, jwtContext, config);
+		validateSignatureAlgorithmWithKey(config, key);
+		validateHeaders(jwtContext, jwtString, config);
+
+		JwtConsumerBuilder consumerBuilder = initializeJwtConsumerBuilderWithValidation(config, jwtClaims, key);
+		JwtConsumer jwtConsumer = consumerBuilder.build();
+		processJwtContextWithConsumer(jwtConsumer, jwtContext);
+	}
+
+	@FFDCIgnore(Exception.class)
+	private void validateHeaders(JwtContext jwtContext, String jwtString, JwtConsumerConfig config) throws Exception {
+
+		if (isRunningBetaMode() && JweHelper.isJwe(jwtString)) {
+			try {
+				String keyManagementKeyAlgorithm = null;
+				// Get keyManagementKeyAlgorithm from server.xml
+				keyManagementKeyAlgorithm = config.getKeyManagementKeyAlgorithm();
+				/**
+				 * If keyManagementKeyAlgorithm from server.xml is null, then
+				 * take the value of keyManagementKeyAlgorithm from
+				 * mpConfigProps
+				 */
+				if (keyManagementKeyAlgorithm == null) {
+					String value = mpConfigProps.get(MpConfigProperties.DECRYPT_KEY_ALGORITHM);
+					if (value != null) {
+						keyManagementKeyAlgorithm = value;
+					}
+				}
+				/**
+				 * If keyManagementKeyAlgorithm is not null, do the following
+				 * check if keyManagementKeyAlgorithm is null (i.e. MP JWT <
+				 * 2.1) skip the check
+				 */
+				if (keyManagementKeyAlgorithm != null) {
+					String[] parts = jwtString.split(("\\."));
```

Comment:
Parsing again seems wrong.  Should be able to call getAlgorithmFromJwtHeader(jwtContext) to get the algorithm, but that is already done in validateAlgorithm so maybe this logic should be there.